include/swift/IDE/CancellableResult.h

```diff
@@ -0,0 +1,148 @@
+//===--- CancellableResult.h ------------------------------------*- C++ -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2021 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef SWIFT_IDE_CANCELLABLE_RESULT_H
+#define SWIFT_IDE_CANCELLABLE_RESULT_H
+
+#include <string>
+
+namespace swift {
+
+namespace ide {
+
+enum class CancellableResultKind { Success, Failure, Cancelled };
+
+/// A result type that can carry one be in one of the following states:
+///  - Success and carry a value of \c ResultType
+///  - Failure and carry an error description
+///  - Cancelled in case the operation that produced the result was cancelled
+///
+/// Essentially this emulates an enum with associated values as follows
+/// \code
+/// enum CancellableResult<ResultType> {
+///   case success(ResultType)
+///   case failure(String)
+///   case cancelled
+/// }
+/// \endcode
+///
+/// The implementation is inspired by llvm::optional_detail::OptionalStorage
+template <typename ResultType>
+class CancellableResult {
+  CancellableResultKind Kind;
+  union {
+    /// If \c Kind == Success, carries the result.
+    ResultType Result;
+    /// If \c Kind == Error, carries the error description.
+    std::string Error;
+    /// If \c Kind == Cancelled, this union is not initialized.
+    char Empty;
+  };
+
+  CancellableResult(ResultType Result)
+      : Kind(CancellableResultKind::Success), Result(Result) {}
+
+  CancellableResult(std::string Error)
+      : Kind(CancellableResultKind::Failure), Error(Error) {}
+
+  explicit CancellableResult()
+      : Kind(CancellableResultKind::Cancelled), Empty() {}
+
+public:
+  CancellableResult(CancellableResult &Other) : Kind(Other.Kind), Empty() {
+    switch (Kind) {
+    case CancellableResultKind::Success:
+      ::new ((void *)std::addressof(Result)) ResultType(Other.Result);
```

Comment:
It’s something to do with deleted copy constructor. IIUC we don’t want to destruct the `Result` in the union because it’s just uninitialized memory and it’s copy contractor might be deleted. I copied the implementation from `llvm::Optional`

https://github.com/apple/llvm-project/blob/e67e2a8d4c65aaa2fa0e41a32630ab5961e0851a/llvm/include/llvm/ADT/Optional.h#L112